vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java

```diff
@@ -2363,6 +2363,83 @@ public HeaderRow appendHeaderRow() {
         return insertInmostColumnLayer(true, false).asHeaderRow();
     }
 
+    /**
+     * Removes the header row from the grid. Note that the default header row
+     * can be removed only if it is the only header row.
+     *
+     * @see #removeAllHeaderRows()
+     * @param headerRow
+     *            the header row to remove
+     * @throws UnsupportedOperationException
+     *             if default row is being removed while there are other header
+     *             rows
+     * @throws NoSuchElementException
+     *             if the header row cannot be found
+     */
+    public void removeHeaderRow(HeaderRow headerRow) {
+        Objects.requireNonNull(headerRow);
+        List<HeaderRow> headerRows = getHeaderRows();
+        if (headerRow.equals(defaultHeaderRow)) {
+            if (headerRows.size() != 1) {
+                throw new UnsupportedOperationException(
+                        "Default header row cannot be removed while there are other header rows.");
+            }
+            removeDefaultHeaderRow();
+        } else {
+            if (!headerRows.contains(headerRow)) {
+                throw new NoSuchElementException(
+                        "Header to remove cannot be found.");
+            }
+            if (getColumnLayers().get(0).equals(headerRow.layer)) {
+                // Bottom layer needs special handling and content manipulation
+                // because it cannot be removed.
+                removeNonDefaultHeaderInBottomLayer(headerRow);
+            } else {
+                removeColumnLayer(headerRow.layer);
+            }
+        }
+    }
+
+    private void removeNonDefaultHeaderInBottomLayer(HeaderRow headerRow) {
+        // Move content from next footer row and remove that layer
+        List<HeaderRow> headerRows = getHeaderRows();
+        HeaderRow nextHeaderRow = headerRows.get(headerRows.size() - 2);
+        ColumnLayer layerToRemove = nextHeaderRow.layer;
+        moveRowContent(nextHeaderRow, headerRow);
+        headerRow.layer.setHeaderRow(nextHeaderRow);
+        if (nextHeaderRow.equals(defaultHeaderRow)) {
+            nextHeaderRow.layer.updateSortingIndicators(true);
+        }
+        layerToRemove.setHeaderRow(null);
+        removeColumnLayer(layerToRemove);
+    }
+
+    /**
+     * Removes all header rows from the grid.
+     *
+     * @see #removeHeaderRow(HeaderRow)
+     */
+    public void removeAllHeaderRows() {
+        List<HeaderRow> headerRows = getHeaderRows();
+        if (headerRows.isEmpty()) {
+            return;
+        }
+        int headerRowCount = headerRows.size();
+        IntStream.range(0, headerRowCount)
+                .mapToObj(index -> headerRows.get(headerRowCount - 1 - index))
+                .filter(headerRow -> !headerRow.equals(defaultHeaderRow))
+                .forEach(this::removeHeaderRow);
+        removeDefaultHeaderRow();
```

Comment:
```suggestion
        getHeaderRows().stream()
                .filter(headerRow -> !headerRow.equals(defaultHeaderRow))
                .forEach(this::removeHeaderRow);
        removeDefaultHeaderRow();
```
This seems to work as well (no reverse order, doesn't matter if there are header rows or not). Do you see any issue with it?

Reply:
There were some issues with the order initial implementation, so I had to impose this order. The current version of `removeHeaderRow` is pretty flexible, and there should be no issues with the order. I think we can remove it.

Reply:
One of the IT's fail with this change. You'll need to reverse the order before processing the header rows:

```java
var headerRows = getHeaderRows();
Collections.reverse(headerRows);
headerRows.stream()...
```

Reply:
Restored the reversing with the suggested cleaner approach.